File: use/neuro-symbolic/pipeline/pipeline.jl
using PyCall

# py_script = read("$(py_script_path)/pipeline.py", String)

# py"""
# $(py_script)
# """
py"""

#!/usr/bin/python3

# py_script_path = "neuro-symbolic/pipeline"
sub_dir = $(py_script_path)
import ast
import os
from wasabi import msg
import pandas as pd
from torch.utils.data import DataLoader
from torch.nn.utils.rnn import pack_padded_sequence, pad_packed_sequence
import torch
from copy import copy
from torch import nn
import numpy as np

class AutoLayer(nn.Module):
		def __init__(self):
				super(AutoLayer, self).__init__()
				pass

		def forward(self, X, attention_mask):
				emb = X
				for layer in self.layers:
						emb = self._get_lstm_out(layer, emb, attention_mask)
				return emb

		def _get_lstm_out(self, layer, emb_input, attention_mask):
				mask = attention_mask.detach().cpu().numpy()
				mask = [elem.index(0) if 0 in elem else len(elem) for elem in mask.tolist()]
				packed_x = pack_padded_sequence(emb_input, mask, batch_first=True, enforce_sorted=False)
				lstm_out,_ = layer(packed_x)
				padded_x,_ = pad_packed_sequence(lstm_out, batch_first=True, total_length=emb_input.shape[1])
				return padded_x

class Encoder(AutoLayer):
		def __init__(self, n_attributes: int, code_size: int):
				super(Encoder, self).__init__()
				
				self.layers = torch.nn.Sequential(
						torch.nn.LSTM(n_attributes, 14, batch_first=True),
						torch.nn.LSTM(14, 7, batch_first=True),
						torch.nn.LSTM(7, code_size, batch_first=True)
						)


class Decoder(AutoLayer):
		def __init__(self, n_attributes: int, code_size: int):
				super(Decoder, self).__init__()
				
				self.layers = nn.Sequential(
						torch.nn.LSTM(code_size, 7, batch_first=True),
						torch.nn.LSTM(7, 14, batch_first=True),
						torch.nn.LSTM(14, n_attributes, batch_first=True),
						)


# basic model
class BasicAutoencoder(torch.nn.Module):
		def __init__(self, n_attributes, n_classes, code_size: int=5):
				super(BasicAutoencoder, self).__init__()
				self.encoder = Encoder(n_attributes, code_size)
				self.decoder = Decoder(n_attributes, code_size)

				self.linear = torch.nn.Linear(code_size, n_classes)
				self.relu = torch.nn.ReLU()
				self.softmax = torch.nn.Softmax(dim=-1)


		def forward(self, X, attention_mask, labels=None):
				encoder_out = self.encoder(X, attention_mask) 

				# if self.training:
				#     cls = self.softmax(self.linear(encoder_out))[:,-1,:]
				#     decoder_out = self.decoder(encoder_out, attention_mask)
				#     return decoder_out, cls
				
				return encoder_out 

import torch
from wasabi import msg
#torch.use_deterministic_algorithms(True)


def get_dataset(dataset):
		if dataset == "fingermovements":
				return "fingermovements", 2, 28
		raise Exception("You must specify a dataset (only \'fingermovements\' is available)")


model_in_use = None    # model initialization

def load_model(dataset, seed):
		global model_in_use
		device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
		dataset, n_classes, n_attributes = get_dataset(dataset)

		path = f"{sub_dir}/assets/checkpoints/model_{dataset}_{1}_seed_{seed}.pt" 
		if not os.path.isfile(path):
				msg.warn(f"No model. Wrong seed or dataset!")
		model_in_use = BasicAutoencoder(n_attributes, n_classes) 
		model_in_use.load_state_dict(torch.load(f"{sub_dir}/assets/checkpoints/model_{dataset}_{1}_seed_{seed}.pt", map_location=device))
		model_in_use = model_in_use.to(device)
		model_in_use.eval()
		model_in_use.training = False

@torch.inference_mode()
def validation(input_serie):
		global model_in_use
		# device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
		
		# with torch.no_grad():
		input_ids = torch.tensor(input_serie,requires_grad = False).unsqueeze(0)#.to(device)
		mask = torch.ones(len(input_serie), requires_grad = False).unsqueeze(0)#.to(device)
		result = model_in_use(input_ids, mask)
		result = result[0,-1,:].squeeze()
		return result.detach().cpu().numpy()#.tolist()

# def get_model(dataset, seed):
# 		device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
# 		dataset, n_classes, n_attributes = get_dataset(dataset)

# 		path = f"{sub_dir}/assets/checkpoints/model_{dataset}_{1}_seed_{seed}.pt" 
# 		if not os.path.isfile(path):
# 				msg.warn(f"No model. Wrong seed or dataset!")
# 		model_in_use = BasicAutoencoder(n_attributes, n_classes) 
# 		model_in_use.load_state_dict(torch.load(f"{sub_dir}/assets/checkpoints/model_{dataset}_{1}_seed_{seed}.pt", map_location=device))
# 		model_in_use = model_in_use.to(device)
# 		model_in_use.eval()
# 		model_in_use.training = False
# 		return model_in_use

# @torch.inference_mode()
# def validation(model_in_use, input_serie):
#     device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
		
#     with torch.no_grad():
#         input_ids = torch.tensor(input_serie).unsqueeze(0).to(device)
#         mask = torch.ones(len(input_serie)).unsqueeze(0).to(device)
#         result = model_in_use(input_ids, mask)
#         result = result[0,-1,:].squeeze()
#         return result.detach().cpu().numpy().tolist()


# @torch.inference_mode()
# def validation(model_in_use, input_serie):
# 		# device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
# 		input_ids = torch.tensor(input_serie, requires_grad = False).unsqueeze(0) #.to(device)
# 		mask = torch.ones(len(input_serie)).unsqueeze(0) #.to(device)
# 		result = model_in_use(input_ids, mask)
# 		result = result[0,-1,:].squeeze()

# 		return result.detach().cpu().numpy()#.tolist()
"""

# py"load_model"("fingermovements", 1)
# py"validation(np.zeros((2,28), dtype=np.float32))"

# model = py"get_model"("fingermovements", 1)
# py"validation"(model, randn(Float32, (2,28)))

File: use/neuro-symbolic/pipeline/pipeline.py
#!/usr/bin/python3
import ast
import os
import pandas as pd
import numpy as np
from model import Seq2Seq
import torch
#torch.use_deterministic_algorithms(True)


model = None    # model initialization

def load_model(dataset, seed, attribute, code, limit):
		""" Carica modello
		
		Args:
				dataset: nome del dataset (tutto minuscolo)
				seed: numero del seed
				attribute: il numero dell'attributo di cui si vuole il modello (si parte da 0)
				limit: 1 se si vogliono considerare tutti i sottointervalli

		Returns:
				None
		"""
		
		_dataset = dataset.lower()
		
		if limit != 1 and limit != 100:
				raise Exception("Limit must be 1 or 100.")
		global model
		device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")

		path = f"assets/checkpoints/{_dataset}_code_{code}/model_code_{code}_{_dataset}_limit_{limit}_seed_{seed}_attr_{attribute}.pt"

		if not os.path.isfile(path):
				raise Exception(f"No model found {path}!")
		model = Seq2Seq(device, code, input_dim = 1, hid_dim = code, dropout=0.25, emb_dim=64) 
		model.load_state_dict(torch.load(path, map_location=device))
		model = model.to(device)



@torch.inference_mode()
def validation(input_serie):
		""" Ritorna la rappresentazione di dimensione 4 per ogni istante temporale

		Args:
				input_serie: str, serie temporale di dimensione 1xn
		
		"""
		global model

		model.eval()
		# input_serie = ast.literal_eval(input_serie)
		device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
		
		with torch.no_grad():
				input_ids = torch.tensor(input_serie).to(device)
				mask = torch.ones(1,input_ids.shape[1])
				result = model(input_ids, None, mask)
				# print(result.numpy().tolist())
				return result.numpy()

def produce_npz_flattened(dataset, seeds, code_size):
	X = np.load(f'../neuro-symbolic-MTSC-data/stump_with_memoization,TestOp_80,{dataset},(false,false,"interval"),missing-X.npy').astype(np.float32)
	n_points, n_attributes, n_instances = X.shape
	print("n_points: ",     n_points)
	print("n_attributes: ", n_attributes)
	print("n_instances: ",  n_instances)

	for seed in seeds:
		flattened_filename = f"flattened_{dataset}-{seed}-{code_size}-X.npy"
		if os.path.isfile(flattened_filename):
			print("Skipping: ", flattened_filename)
		else:
			flattened_array = np.ndarray((code_size, n_attributes, n_instances))
			for attribute in range(n_attributes):
				print(f"Attribute {attribute}/{n_attributes}")
				load_model(dataset, seed, attribute, code_size, 100)
				for instance in range(n_instances):
					# print(X[:,attribute,instance].shape)
					# print(np.expand_dims(X[:,attribute,instance], axis=0).transpose().shape)
					# res = validation(np.expand_dims(X[:,attribute,instance], axis=0).transpose())
					res = validation([X[:,attribute,instance].tolist()])
					# print(res)
					flattened_array[:,attribute,instance] = res
			# print(flattened_array)
			# print(flattened_array.shape)
			print(f"Saving {flattened_filename}...")
			np.save(flattened_filename, flattened_array)

def produce_npz_fmd(dataset, seeds, code_size):
	X = np.load(f'../neuro-symbolic-MTSC-data/stump_with_memoization,TestOp_80,{dataset},(false,false,"interval"),missing-X.npy').astype(np.float32)
	n_points, n_attributes, n_instances = X.shape
	print("n_points: ",     n_points)
	print("n_attributes: ", n_attributes)
	print("n_instances: ",  n_instances)

	for seed in seeds:
		fmd_filename = f"fmd_{dataset}-{seed}-{code_size}-X.npy"
		if os.path.isfile(fmd_filename):
			print("Skipping: ", fmd_filename)
		else:
			fmd_array = np.ndarray((code_size,n_points,n_points+1,n_attributes, n_instances))
			for attribute in range(n_attributes):
				print(f"Attribute {attribute}/{n_attributes}")
				load_model(dataset, seed, attribute, code_size, 1)
				for x in range(n_points):
					for y in range(x+1,n_points+1):
						for instance in range(n_instances):
							# print(str(x)+":"+str(y))
							# print(X[x:y,attribute,instance].shape)
							# print(np.expand_dims(X[x:y,attribute,instance], axis=0).shape)
							# print(np.expand_dims(X[x:y,attribute,instance], axis=0).transpose().shape)
							# res = validation(np.expand_dims(X[x:y,attribute,instance], axis=0).transpose())
							# print(res)
							# res = validation(np.expand_dims(X[x:y,attribute,instance], axis=0).transpose())
							res = validation([X[x:y,attribute,instance].tolist()])
							# print(res)
							fmd_array[:,x,y,attribute, instance] = res
			print(f"Saving {fmd_filename}...")
			np.save(fmd_filename, fmd_array)
			print()

################################################################################
################################################################################
################################################################################

# load_model("Libras", 1, 0, 1)
print("Testing a RacketSports model...")
load_model("RacketSports", 1, 0, 1, 1)
# load_model("RacketSports", 1, 0, 2, 1)
print("Result:", validation([[1,2,3,1,2,3,4.]]))
print("Result:", validation([[1],[2],[3],[1],[2],[3],[4.]]))
print()
# load_model("RacketSports", 1, 0, 4, 1)
print("Testing a Libras model...")
load_model("Libras", 1, 0, 1, 1)
print("Result:", validation([[1.]]))
print()


################################################################################
################################################################################
################################################################################

for code_size in [1]: # ,2,4]:
	for dataset in ["RacketSports", "Libras", "NATOPS"]:
		seeds = [1,2,3,4,5]
		produce_npz_flattened(dataset, seeds, code_size)

for code_size in [1]: # ,2,4]:
	for dataset in ["RacketSports", "Libras", "NATOPS"]:
		seeds = [1,2,3,4,5]
		produce_npz_fmd(dataset, seeds, code_size)
